# Genetics-Workshop-Mexico-LAI-2024

## **üåê Overview**

This workshop provides a **comprehensive guide** to performing **local ancestry inference** using **DNAnexus** and the **RFMix** toolset. The process is divided into two main parts:

### **üîß Setting Up DNAnexus**
- **Configure** a stable working environment through snapshots.

### **üöÄ Ancestry Pipeline Execution**
- After setting up DNAnexus, **run the ancestry pipeline** to process phased genetic data.
- **Execute RFMix** for local ancestry inference.
- **Visualize** the results.

By following this guide, you'll gain a **thorough understanding** of the entire process, from **setting up the environment** in DNAnexus to **executing the analysis pipeline** and **interpreting the results** generated by the `phased_rfmix_dn.sh` script.

---

## **üìë Outline**

### **[1Ô∏è‚É£ Part 1: Running DNAnexus](#running-dnanexus)**

1. **[Snapshot of DNAnexus Project](#set-up-snapshot)**
   - Verify and use the snapshot as needed.


### **[2Ô∏è‚É£ Part 2: Ancestry Pipeline](#ancestry-pipeline)**

1. **[Pipeline Description](#pipeline-description)**
   - Overview of the pipeline, which processes phased genetic data, runs RFMix for local ancestry inference, collapses RFMix output into BED files, and generates karyograms and global ancestry proportions.

2. **[Pipeline Steps](#pipeline-steps)**
   - **1. Infer Local Ancestry**
     - Execute RFMix to infer local ancestry for each individual.
   - **2.1 Collapse Inferred Data**
     - Collapse RFMix output into BED files, plot karyograms, and estimate global ancestry proportions.

---

## Running DNAnexus

### **Step 1: Navigate to and Use the Snapshot**

1. **Navigate to the Snapshot**:
   - To work within the snapshot, navigate to it using the following command.

   ```bash
   dx cd $DX_PROJECT_CONTEXT_ID:snapshot-my_rfMix_workflow_snapshot
   ```

   - Replace `$DX_PROJECT_CONTEXT_ID` with your project‚Äôs ID and `my_rfMix_workflow_snapshot` with the snapshot name.

2. **Run Workflows from the Snapshot**:
   - You can run your applets or workflows from within this snapshot as if you were working in the main project environment.

   ```bash
   dx run your_applet_or_workflow_name --input_vcf=input.vcf --genetic_map=genetic_map.txt --classes_txt=classes.txt --sample_map=sample_map.txt
   ```

---

## Ancestry Pipeline

### **Description**
This pipeline is designed to perform local ancestry inference on phased genetic data. The pipeline takes phased data formats for RFMix, runs the local ancestry inference using RFMix, collapses the RFMix output into BED files, plots karyograms, and estimates global ancestry proportions.

**Key References:**
- **1000 Genomes Data:** [Link](https://www.internationalgenome.org/data/)
- **Human Genome Diversity Project (HGDP):** [Link](https://www.internationalgenome.org/data-portal/data-collection/hgdp)

Slides from a tutorial are available in this repository.

---

### **Pipeline Map**

You should have phased data in a format that aligns with RFMix specifications.

#### **1. Infer Local Ancestry**
  * **Run RFMix:** This is the core step of the pipeline where local ancestry is inferred for each individual. RFMix uses reference populations to assign ancestry to each segment of the genome, providing a detailed picture of an individual's ancestry composition.

#### **2.1 Collapse Inferred Data**
  * **Collapse RFMix Output into TRACTS-Compatible BED Files:** The output from RFMix is processed to generate BED files that are compatible with the TRACTS software, which is used for further analysis and visualization.
  * **Posthoc BED File Filter (OPTIONAL):** An optional step where the BED files can be filtered to remove low-confidence regions or other unwanted data.
  * **Plot Ancestry Karyograms:** Visualization of the inferred ancestry along the genome, typically displayed in karyogram format.
  * **Estimate Global Ancestry Proportions from Local Ancestry Inference:** Summarize the local ancestry results to estimate overall ancestry proportions for each individual.

---

### **0. Phase**

#### **Overview**

This step involves running a phasing algorithm (e.g., SHAPEIT4) that arranges the genetic data in a phased format. Phasing is critical as it allows for the correct interpretation of genetic variants by distinguishing between the two sets of chromosomes an individual inherits from their parents.

Here's an example of how to run SHAPEIT4 across all chromosomes:

```bash
for i in {1..22}; 
do shapeit4 \
  --input "$data_dir/input${i}.vcf" \
  --map "$data_dir/genetic_map${i}.txt" \
  --region 1-22 \
  --output "$output_dir/phased_output${i}.vcf" \
  --log "$output_dir/shapeit4${i}.log"; done
```

Phasing should be parallelized across chromosomes and can be run with PLINK files or VCF files.

---

### **1. Infer Local Ancestry**

#### **Run RFMix**

RFMix is a powerful tool that uses phased genetic data and reference panels to infer local ancestry. In this step, the tool assigns ancestry labels to different segments of the genome, giving a detailed breakdown of where different ancestries are located on each chromosome.

Here‚Äôs how to run RFMix:

```bash
for i in {1..22}; do \
rfmix \
  --input-classes "$data_dir/classes.txt" \
  --input-genetic-map "$data_dir/genetic_map.txt" \
  --input-sample-map "$data_dir/sample_map.txt" \
  --input-vcf "$output_dir/phased_output.vcf" \
  --output-dir "$output_dir/rfmix_output${i}"; done
```

---


#### **Plot Ancestry Karyograms**

A karyogram is a visual representation of the chromosomes and the inferred ancestry across them. This step uses a script to plot the karyograms based on the BED files generated earlier.

```bash
IND='HG02481'; python plot_karyogram.py \
--bed_a ${IND}_A.bed \
--bed_b ${IND}_B.bed \
--ind ${IND} \
--out ${IND}.png
```


---

#### **Estimate Global Ancestry Proportions from Local Ancestry Inference**

The final step is to estimate global ancestry proportions, which are summarized from the local ancestry data. This comparison can be useful for assessing agreement with other methods like ADMIXTURE.

```bash
Q files, script to make global admixture
```



The `ind_list` input contains the individual IDs for summarizing the output. The `--pops` option specifies the populations for which to estimate global ancestry proportions. This allows for the exclusion of unknown (UNK) tracts from the estimation. An example output file is attached.


### Summary of Main Components:

1. **Phasing (Step 0)**: Phasing aligns genetic variants on homologous chromosomes, creating the foundation for accurate local ancestry inference. This step uses algorithms like SHAPEIT4, which arrange the sequences in a way that differentiates between the two sets of chromosomes inherited from the parents.

2. **Local Ancestry Inference (Step 1)**: This step uses RFMix to assign ancestry to segments of the genome. By comparing an individual‚Äôs phased genetic data to reference populations, RFMix provides a detailed breakdown of ancestry across the genome. This information is crucial for understanding the genetic makeup of an individual and identifying regions of the genome that come from specific ancestral populations.

